File: library/src/main/kotlin/results/EncodeResult.kt
package results

sealed class EncodeResult {
    data class Success(
        val byteArray: ByteArray
    ) : EncodeResult() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Success

            return byteArray.contentEquals(other.byteArray)
        }

        override fun hashCode(): Int {
            return byteArray.contentHashCode()
        }
    }

    data class Fatal(val error: Any? = null) : EncodeResult()
}

File: python-library/results/encode_result.py
from dataclasses import dataclass
from typing import Any, Optional


@dataclass
class EncodeResult:
    @dataclass
    class Success:
        byte_array: bytes

        def __eq__(self, other) -> bool:
            if not isinstance(other, EncodeResult.Success):
                return False
            return self.byte_array == other.byte_array

        def __hash__(self) -> int:
            return hash(self.byte_array)

    @dataclass
    class Fatal:
        """Failed encoding result"""
        error: Optional[Any] = None